api/desecapi/pdns_change_tracker.py

```diff
@@ -0,0 +1,228 @@
+                                rrset__type=type_,
+                                rrset__subname=subname)
+                        ]
+                    }
+                    for type_, subname in (additions | modifications) - deletions
+                ] + [
+                    {
+                        'name': '.'.join(filter(None, [subname, domain_name])) + '.',
+                        'type': type_,
+                        'changetype': 'REPLACE',
+                        'records': []
+                    }
+                    for type_, subname in deletions
+                ]
+            }
+
+            if data['rrsets']:
+                _pdns_patch('/zones/' + pdns_id, data)
+
+            if data['rrsets'] or domain_name in self._domain_additions:
+                _pdns_put('/zones/%s/notify' % pdns_id)
+
+    @staticmethod
+    def _get_ttl_or_zero(domain_name, type_, subname):
+        result = RRset.objects.filter(domain__name=domain_name, type=type_, subname=subname)
+        return result[0].ttl if result else 0
+
+    def _update_rr_set(self, rr_set: RRset, deleted=False, created=False):
+        if self._rr_set_modifications.get(rr_set.domain.name, None) is None:
+            self._rr_set_additions[rr_set.domain.name] = set()
+            self._rr_set_modifications[rr_set.domain.name] = set()
+            self._rr_set_deletions[rr_set.domain.name] = set()
+
+        additions = self._rr_set_additions[rr_set.domain.name]
+        modifications = self._rr_set_modifications[rr_set.domain.name]
+        deletions = self._rr_set_deletions[rr_set.domain.name]
+
+        item = (rr_set.type, rr_set.subname)
+        if created:
+            additions.add(item)
+            assert item not in modifications
+            deletions.discard(item)
+        elif deleted:
+            if item in additions:
+                additions.remove(item)
+                modifications.discard(item)
+                # no change to deletions
+            else:
+                # item not in additions
+                modifications.discard(item)
+                deletions.add(item)
+        elif not created and not deleted:
+            # we don't care if item was created or not
+            modifications.add(item)
+            assert item not in deletions
+        else:
+            raise ValueError('An RR set cannot be created and deleted at the same time.')
+
+    def _update_domain(self, domain: Domain, created=False, deleted=False):
+        if not created and not deleted:
+            raise UnsupportedChangeException()  # TODO can this be relaxed to just no-op?
+
+        name = domain.name
+        additions = self._domain_additions
+        deletions = self._domain_deletions
+
+        if created and deleted:
+            raise ValueError('A domain set cannot be created and deleted at the same time.')
+
+        if created:
+            if name in deletions:
+                deletions.remove(name)
```

Comment:
I am not sure I understand the situation in which `created` can be `True` and the domain is already up for deletion. Is this the case where a domain is deleted and immediately recreated? If so, don't we have to actually delete the domain to get rid of all records, and also to generate new keys?

If this is something entirely different, sorry. :)

Reply:
Let's say your code does the following:

    with PDNSChangeTracker():
        d = Domain.create(name='foobar.example.com',...)  # INSERT query
    # we issue a create zone request to nslord

    with PDNSChangeTracker():
        d.delete()  # DELETE query
        # note that at this point, still no query to pdns was made, but one deletion is pending
        new_d = Domain.create(name='foobar.example.com', ...)
        # at this point, we removed the pending query because `name in creations` is true
    # we issue no request

This looks like things can go wrong, as the new domain would be expected to be empty. However, I believe this will be taken care of by the removal of any existing `RRsets`. I'll add tests for those cases.